src/ucp/core/ucp_listener.c

```diff
@@ -90,63 +104,83 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
 }
 
-static ucs_status_t ucp_listener_conn_request_callback(void *arg,
-                                                       const void *conn_priv_data,
-                                                       size_t length)
+ucs_status_t
+ucp_listener_ep_create(ucp_listener_h                    listener,
+                       const ucp_wireup_sockaddr_priv_t *client_data,
+                       ucp_listener_accept_t            *accept)
+{
+    ucp_ep_h     ep;
+    ucs_status_t status = ucp_ep_create_accept(listener->wiface.worker,
+                                               client_data, &ep);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    accept->listener = listener;
+    accept->is_ep    = 1;
+    accept->ep       = ep;
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucp_listener_ep_addr_create(ucp_listener_h listener, void *id,
+                            const ucp_wireup_sockaddr_priv_t *client_data,
+                            size_t length, ucp_listener_accept_t *accept)
+{
+    accept->listener    = listener;
+    accept->is_ep       = 0;
+    accept->addr        = ucs_malloc(ucs_offsetof(ucp_ep_address_t, priv_addr) +
+                                     length, "accept ep addr");
+    if (accept->addr == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    accept->addr->listener    = listener;
+    accept->addr->conn_req_id = id;
+    memcpy(&accept->addr->priv_addr, client_data, length);
+
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucp_listener_conn_request_callback(void *arg, void *id,
+                                   const void *conn_priv_data, size_t length)
 {
     const ucp_wireup_sockaddr_priv_t *client_data = conn_priv_data;
     ucp_listener_h listener                       = arg;
-    ucp_unpacked_address_t remote_address;
     ucp_listener_accept_t *accept;
     uct_worker_cb_id_t prog_id;
-    ucp_ep_params_t params;
     ucs_status_t status;
-    ucp_ep_h ep;
 
     ucs_trace("listener %p: got connection request", listener);
 
-    params.field_mask = UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
-    params.err_mode   = client_data->err_mode;
-
-    status = ucp_address_unpack(client_data + 1, &remote_address);
-    if (status != UCS_OK) {
-        goto err;
+    /* Defer wireup init and user's callback to be invoked from the main thread */
+    accept = ucs_malloc(sizeof(*accept), "ucp_listener_accept");
+    if (accept == NULL) {
+        ucs_error("failed to allocate listener accept context");
+        return UCS_ERR_NO_MEMORY;
     }
 
-    if (client_data->is_full_addr) {
-        /* create endpoint to the worker address we got in the private data */
-        status = ucp_ep_create_to_worker_addr(listener->wiface.worker, &params,
-                                              &remote_address, UCP_EP_CREATE_AM_LANE,
-                                              "listener", &ep);
-        if (status != UCS_OK) {
-            goto err_free_address;
+    if (listener->cb) {
```

Comment:
shouldn't the rest of this function be thread safe to protect the worker?

Reply:
Seems like yes because 

1. 
 https://github.com/openucx/ucx/blob/706ffc57d5587e2c012a6feced2a5aec36171edb/src/ucp/core/ucp_listener.c#L230

2. 
 https://github.com/openucx/ucx/blob/706ffc57d5587e2c012a6feced2a5aec36171edb/src/uct/ib/rdmacm/rdmacm_iface.c#L140

and this is not a regression, right?